glyphs-reader/src/propagate_anchors.rs

```diff
@@ -0,0 +1,883 @@
+    pub fn propagate_all_anchors(&mut self) {
+        propagate_all_anchors_impl(&mut self.glyphs);
+    }
+}
+
+// the actual implementation: it's easier to test a free fn
+fn propagate_all_anchors_impl(glyphs: &mut BTreeMap<SmolStr, Glyph>) {
+    // the reference implementation does this recursively, but we opt to
+    // implement it by pre-sorting the work to ensure we always process components
+    // first.
+    let todo = depth_sorted_composite_glyphs(glyphs);
+    let mut num_base_glyphs = HashMap::new();
+    // NOTE: there's an important detail here, which is that we need to call the
+    // 'anchors_traversing_components' function on each glyph, and save the returned
+    // anchors, but we only *set* those anchors on glyphs that have components.
+    // to make this work, we write the anchors to a separate data structure, and
+    // then only update the actual glyphs after we've done all the work.
+    let mut all_anchors = HashMap::new();
+    for name in todo {
+        // temporarily remove the glyph so we can mess with it without borrowing all glyphs
+        let glyph = glyphs.get(&name).unwrap();
+        // we work with indices here to also get around borrowck, because
+        // if we borrow the layers in the loop we can't set the anchors in the loop body
+        for layer in &glyph.layers {
+            let anchors = anchors_traversing_components(
+                glyph,
+                layer,
+                glyphs,
+                &all_anchors,
+                &mut num_base_glyphs,
+            );
+            if anchors.len() != layer.anchors.len() {
+                debug_new_anchors(&anchors, glyph, layer);
+            }
+            all_anchors.entry(name.clone()).or_default().push(anchors);
+        }
+    }
+
+    // finally update our glyphs with the new anchors, where appropriate
+    for (name, layers) in all_anchors {
+        let glyph = glyphs.get_mut(&name).unwrap();
+        if glyph.has_components() {
+            assert_eq!(layers.len(), glyph.layers.len());
+            for (i, layer_anchors) in layers.into_iter().enumerate() {
+                glyph.layers[i].anchors = layer_anchors;
+            }
+        }
+    }
+}
+
+fn debug_new_anchors(anchors: &[Anchor], glyph: &Glyph, layer: &Layer) {
+    if !log::log_enabled!(log::Level::Info) {
+        return;
+    }
+    let prev_names: Vec<_> = layer.anchors.iter().map(|a| &a.name).collect();
+    let new_names: Vec<_> = anchors.iter().map(|a| &a.name).collect();
+    log::info!(
+        "propagated anchors for ('{}': {prev_names:?} -> {new_names:?}",
+        glyph.name,
+    );
+}
+
+/// Return the anchors for this glyph, including anchors from components
+fn anchors_traversing_components<'a>(
+    glyph: &'a Glyph,
+    layer: &'a Layer,
+    glyphs: &BTreeMap<SmolStr, Glyph>,
+    // map of glyph -> anchors, per-layer, updated as we do each glyph
+    visited_anchors: &HashMap<SmolStr, Vec<Vec<Anchor>>>,
```

Comment:
`visited` in a graph traversal led me to expect a fairly specific pattern of use (if already visited, return that answer, otherwise compute and record) that this doesn't follow. I wonder if it's worth renaming to just `anchors`